### ISSUE TITLE
Setting up inflow condition in bounded domain with buoyancy as tracer

### ISSUE THREAD

**Issue by raphaelouillon** [CliMA/Oceananigans.jl#789]
Hi all, I am trying to setup a DNS in a triply bounded domain with inflow conditions on the west wall. All boundary conditions are as default except for the bottom boundary (no-slip), west wall (inflow) and right wall (outflow). I tried setting up the boundaries as
```
ubcs = UVelocityBoundaryConditions(grid, bottom = BoundaryCondition(Value, 0), west = BoundaryCondition(Value,0.1), east = BoundaryCondition(Gradient,0))
vbcs = VVelocityBoundaryConditions(grid, bottom = BoundaryCondition(Value, 0))
bbcs = TracerBoundaryConditions(grid, west = BoundaryCondition(Value,0))
```
The tracer boundary condition is fixed to 0 on the west wall. I am getting the following error when running the simulation:

ERROR: LoadError: MethodError: no method matching fill_west_halo!(::OffsetArrays.OffsetArray{Float64,3,Array{Float64,3}}, ::BoundaryCondition{Value,Float64}, ::CPU, ::RegularCartesianGrid{Float64,Bounded,Bounded,Bounded,OffsetArrays.OffsetArray{Float64,3,Base.ReshapedArray{Float64,3,StepRangeLen{Float64,Base.TwicePrecision{Float64},Base.TwicePrecision{Float64}},Tuple{}}}}, ::Clock{Float64}, ::NamedTuple{(:velocities, :tracers, :diffusivities),Tuple{NamedTuple{(:u, :v, :w),Tuple{OffsetArrays.OffsetArray{Float64,3,Array{Float64,3}},OffsetArrays.OffsetArray{Float64,3,Array{Float64,3}},OffsetArrays.OffsetArray{Float64,3,Array{Float64,3}}}},NamedTuple{(:b,),Tuple{OffsetArrays.OffsetArray{Float64,3,Array{Float64,3}}}},Nothing}})

I now wonder if it is currently possible to have inflow conditions with buoyancy, or if this is a boundary condition that is not implemented. Are inflow conditions incompatible with buoyancy or is the error unrelated? Thanks very much for your help! Entire script below:

```
# # Lock-release gravity current example

using Oceananigans, Oceananigans.Grids, Printf, SpecialFunctions, Plots


# ## Physical and numerical parameters
#
# First, we pick a resolution and domain size,

Nx = 256
Ny = 256 # x resolution
Nz = 32  # z resolution
Lx = 10
Ly = 12  # domain extent
Lz = 1   # vertical domain extent
#y₀ = 2   # Initial source position
#δᴸ = 0.1 # thickness of lock transition

Re = 2000
Pe = Re
iRe = 1/Re
iPe = 1/Pe
nothing # hide


# ## Build the grid
#
# We use a domain that is bounded in y (north and south) and z (top and bottom),
# and periodic in x.
#
topology = (Bounded, Bounded, Bounded)

grid = RegularCartesianGrid(size=(Nx, Ny, Nz), extent=(Lx, Ly, Lz), topology=topology)

# No-slip bottom, free-slip top and side walls through default boundary conditions
ubcs = UVelocityBoundaryConditions(grid, bottom = BoundaryCondition(Value, 0), west = BoundaryCondition(Value,0.1), east = BoundaryCondition(Gradient,0.0))
vbcs = VVelocityBoundaryConditions(grid, bottom = BoundaryCondition(Value, 0))
bbcs = TracerBoundaryConditions(grid, west = BoundaryCondition(Value,0.0))
nothing # hide

# ## Model instantiation and initial condition
#
# We use a constant viscosity and diffusivity.
# Set forcing
#fun_forcing(x, y, z, t) = (0.5*(1-erf(abs(sqrt((x-xc)^2+(y-yc)^2+(z-zc)^2)-R))))
fun_dist_to_source(x,y,z,t,a,y₀) = sqrt((x-Lx/2)^2+(y-y₀-a*t)^2+(z+Lz/2)^2)
fun_forcing(x, y, z, t,p) = -0.5*(1-erf((fun_dist_to_source(x,y,z,t,p.a,p.y₀)-p.R)/p.δᴸ))

b_forcing = SimpleForcing(fun_forcing, parameters=(a=0.5,y₀=2,R=0.5,δᴸ=0.1))

#parameterized_forcing(x, y, z, t, p) = p.μ * exp(z/p.λ) * cos(p.ω*t)
#v_forcing = SimpleForcing(parameterized_forcing, parameters=(μ=42, λ=0.1, ω=π))
forcing = ModelForcing(b=b_forcing)

model = IncompressibleModel(
                   grid = grid,
           architecture = CPU(),
    boundary_conditions = (u=ubcs, v=vbcs, b=bbcs),
                closure = ConstantIsotropicDiffusivity(ν=iRe, κ=iPe),
                tracers = :b,
               buoyancy = BuoyancyTracer(),
               forcing  = forcing
)

# Set initial condition
b₀(x, y, z) = 0.0#-0.5 * (1 - erf((y - yᴸ) / δᴸ))


set!(model, b=b₀)
nothing # hide

### Progress diagnostic function
using Oceananigans.Diagnostics, Printf

umax = FieldMaximum(abs, model.velocities.u)
vmax = FieldMaximum(abs, model.velocities.v)
wmax = FieldMaximum(abs, model.velocities.w)

wall_clock = time_ns()

function print_progress(simulation)
model = simulation.model

# Print a progress message
msg = @sprintf("i: %04d, t: %s, Δt: %s, umax = (%.1e, %.1e, %.1e) ms⁻¹, wall time: %s\n",
model.clock.iteration,
prettytime(model.clock.time),
prettytime(wizard.Δt),
umax(), vmax(), wmax(),
prettytime(1e-9 * (time_ns() - wall_clock))
)

@info msg

return nothing
end

# Time step wizard based on CFL condition for AB (CFL<=0.3)
wizard = TimeStepWizard(cfl=0.3, Δt=0.01, max_change=1.2, max_Δt=0.1)

nothing # hide

simulation = Simulation(model,progress_frequency = 100, Δt = wizard,stop_time = 20,progress = print_progress)

nothing # hide



# Output
#
# We set up an output writer for the simulation that saves all velocity fields,
# tracer fields, and the subgrid turbulent diffusivity every 2 minutes.

using Oceananigans.OutputWriters

field_outputs = FieldOutputs(merge(model.velocities, model.tracers))

simulation.output_writers[:fields] = JLD2OutputWriter(model, field_outputs,
                                                      interval = 0.1,
                                                        prefix = "continuous_release_3D_background_current",
                                                         force = true)

nothing # hide


run!(simulation)
nothing # hide
```

**Comment by glwagner** [comment]
@raphaelouillon to prescribed in-flow, you need to use `NormalFlow` for the wall-normal velocity component, rather than `Value`. This feature has not been tested so keep us updated on your progress!

**Comment by glwagner** [comment]
To get this simulation running in current Oceananigans, I think you should put a sponge layer near the outflow that relaxes the velocity field to the same velocity as the inflow, and then set the outflow as identical to the inflow. In other words, change the `ubcs` to

```julia
ubcs = UVelocityBoundaryConditions(grid, bottom = BoundaryCondition(Value, 0), west = BoundaryCondition(NormalFlow, 0.1), east = BoundaryCondition(NormalFlow, 0.1))
```

For the relaxation you can use the `Relaxation` abstraction:

https://github.com/CliMA/Oceananigans.jl/blob/1723136b1249332eb1ca4ff3adfad3c38017c452/src/Forcing/relaxation.jl#L23-L40

I think something like

```julia
u_forcing = Relaxation(; mask=GaussianMask{:x}(grid.Lx, grid.Lx/10), target=0.1)
```

should work.

In the longer term, we can potentially implement a new boundary condition type...

**Comment by glwagner** [comment]
See the docs for `GaussianMask` here:

https://github.com/CliMA/Oceananigans.jl/blob/1723136b1249332eb1ca4ff3adfad3c38017c452/src/Forcing/relaxation.jl#L47-L60

**Comment by glwagner** [comment]
Once #790 is merged you'll be able to write

```julia
u_forcing = Relaxation(; rate=1/60, mask=GaussianMask{:x}(grid.Lx, grid.Lx/10), target=0.1)
```

**Comment by raphaelouillon** [comment]
Thanks very much @glwagner. I tried what you suggested but it looks like NormalFlux is not defined in the version I am using. I tried adding NormalFlux and NormalFluxBoundaryCondition to the export in Oceananigans.jl but does not look like it solved it.

**Comment by glwagner** [comment]
try

```julia
using Oceananigans.BoundaryConditions: NormalFlux
```

**Comment by raphaelouillon** [comment]
Thanks again @glwagner, unfortunately I still get the same error (ERROR: LoadError: UndefVarError: Relaxation not defined). I have been trying to figure out why, so far no success. Will keep you updated, thanks again!

**Comment by ali-ramadhan** [comment]
Sorry just approved PR #790 and merged it into master so @glwagner's suggestion should work now.

@raphaelouillon Which version of Oceananigans are you currently using? You can check by typing `] status` at the Julia REPL in package mode (which you enter by pressing `]`).

You might need to use/depend on the `#master` branch of Oceananigans.jl. If you're not sure how to do so, I think you just type `] add Oceananigans#master` in the Julia REPL in package mode.

Let me know if you're still hitting issues!

**Comment by raphaelouillon** [comment]
Thanks very much @ali-ramadhan. I am now using version "Oceananigans v0.30.0 #master (https://github.com/CliMA/Oceananigans.jl.git)". 

I only get an error when calling 

`u_forcing = Relaxation(; rate=1/60, mask=GaussianMask{:x}(grid.Lx, grid.Lx/10), target=0.1)`

which returns ERROR: LoadError: MethodError: no method matching GaussianMask{:x,T} where T(::Float64, ::Float64)

**Comment by ali-ramadhan** [comment]
Ah sorry about that @raphaelouillon! Working from master can be kinda risky and we still need to update the docs...

Not sure why `GaussianMask{:x}(grid.Lx, grid.Lx/10)` would error. Probably deserves its own issue and tests.

Can you try `GaussianMask{:x}(center=grid.Lx, width=grid.Lx/10)` instead? Seems to work for me on Julia 1.4.2.

```julia
julia> using Oceananigans, Oceananigans.Forcing
julia> grid= RegularCartesianGrid(size=(64, 64, 64), extent=(1, 1, 1))
julia> u_forcing = Relaxation(; rate=1/60, mask=GaussianMask{:x}(center=grid.Lx, width=grid.Lx/10), target=0.1)
SimpleForcing{Oceananigans.Grids.Cell,Oceananigans.Grids.Cell,Oceananigans.Grids.Cell,true,Nothing,:tracer,:tracers,Oceananigans.Forcing.RelaxingFunction{Float64,GaussianMask{:x,Float64},Float64}}(Oceananigans.Forcing.RelaxingFunction{Float64,GaussianMask{:x,Float64},Float64}(0.016666666666666666, GaussianMask{:x,Float64}(1.0, 0.1), 0.1), nothing, true)
```

**Comment by raphaelouillon** [comment]
Thanks a lot @ali-ramadhan, that solved it for me!

**Comment by glwagner** [comment]
Ah apologies --- my syntax was wrong. The docstring is incorrect there, as you can see from the inner constructor for `GaussianMask`.

**Comment by glwagner** [comment]
@raphaelouillon is this issue resolved?
